### PR TITLE
feat: asciidoctor plugin

### PR DESCRIPTION
**Pull request by nicorikken**
A basic tasks for processing asciidoc files using AsciidoctorJ. This implementation is bare-bones, it does not cover Asciidoctor extensions and does not relate asciidoc attributes with Perun meta.
The page title is not rendered by the Asciidoctor task, to mimic the behavior of the markdown extension.

### CODE REVIEW

**Comment by nicorikken** [comment]
For the backstory of this PR, please read https://github.com/hashobject/perun/pull/100

**Comment by nicorikken** [comment]
Any feedback?

**Comment by bhagany** [comment]
Hi Nico!  I'm so glad you've found time to work on this.  I apologize - you've caught me in the middle of a job transition (today is my second day) and I haven't had time to do a proper review.  I should be able to do so this weekend, though.  Sorry for the delay!

**Comment by nicorikken** [comment]
@bhagany Thanks for letting me know. I'm excited to have something mergeable, so I'm eager to get it reviewed, that's all. But certainly it can wait some more, no problem.

**Review by podviaznikov — COMMENTED** [review]
code looks good to me. Waiting for @bhagany review before merging

File: src/io/perun.clj
Comment (bhagany) [inline src/io/perun.clj:504]: Since Asciidoc can use YAML front-matter itself, you might consider passing `:keep-yaml true` to `yaml-metadata`.  That way, the YAML will still be present in the file when the `asciidoctor*` task sees it.

File: src/io/perun/asciidoctor.clj
Comment (bhagany) [inline src/io/perun/asciidoctor.clj]: This might not make much difference, but rather than creating the Asciidoctor interface on each call to `asciidoctor-to-html`, would it be more efficient to say, `(def asciidcotor (Asciidoctor$Factory/create ""))` as a top-level form, and then reuse the same object?

File: test/io/perun_test.clj
Comment (bhagany) [inline test/io/perun_test.clj:428]: I like that you're testing different parsing modes

**Comment by bhagany** [comment]
I'm very pleased with how clean this implementation is.  I think it's a good base on which more Asciidoctor features can be added over time.  Thanks!

File: src/io/perun.clj
Comment (nicorikken) [reply src/io/perun.clj:504]: I agree, but the issue arises that the default meta keys used in Perun and !{Asciidoctor](http://asciidoctor.org/docs/user-manual/#attribute-catalog) are different. I was thinking of introducing a translation layer in a next version which contains the necessary translation layer, as I stated in the currently last comment on https://github.com/hashobject/perun/pull/100 For now I could indeed also just pass it through. What do you think, considering my comments? Both are fine to me.

File: src/io/perun/asciidoctor.clj
Comment (nicorikken) [reply src/io/perun/asciidoctor.clj]: Yes, but this fails the moment that environment-specific parameters are introduced like libraries. The issue of creating new containers came up on https://github.com/hashobject/perun/pull/100 and does hurt performance for large number of conversions.
I guess not creating new containers does limit the conversion process to a single core?
Probably the converter-specific settings have to be raised to the task level, so that a different container setting requires a different task to be created. Then inside the task the same converter (factory) can be used.
I will refactor this.

File: test/io/perun_test.clj
Comment (nicorikken) [reply test/io/perun_test.clj:428]: Then again I couldn't get the tests to fail based on content, so perhaps I'm using the framework incorrectly.

**Comment by nicorikken** [comment]
@bhagany Let me know what you think about the options for dealing with YAML, and if I'm missing out om some testing features in the testing framework. I'll refactor the initiation of new JRuby containers.

File: src/io/perun/asciidoctor.clj
Comment (MartyGentillon) [inline src/io/perun/asciidoctor.clj:1]: See above.

File: src/io/perun/asciidoctor.clj
Comment (MartyGentillon) [inline src/io/perun/asciidoctor.clj]: See above.

File: src/io/perun.clj
Comment (MartyGentillon) [inline src/io/perun.clj:494]: By naming the task after the library you are using, you bind the implementation to the interface. A better task name would be ascidoc.

**Review by MartyGentillon — COMMENTED** [review]
Nice work.  A few thoughts.

**Comment by nicorikken** [comment]
@MartyGentillon I see your point of view, but then again, there is also an outdated [Python reference implementation of Asciidoc](http://www.methods.co.nz/asciidoc/) which is quite outdated  Small parts of syntax in Asciidoctor are still improving, to end up at the end-goal of an 'unidoc' syntax. Also Asciidoctor has some Markdown support, and has the ability to load some plugins, like the terrific [asciidoctor-diagram](https://github.com/asciidoctor/asciidoctor-diagram). So while I see where you are coming from, I'd rather leave it at the tool, for clarity.

**Comment by MartyGentillon** [comment]
That seems like a fairly good argument to me.  I wonder if it makes sense
to add aliases for those who don't care what Asciidoc interpreter they
use.  It would give the best of both worlds, people woo don't care won't be
broken by a parser change, while those who do can use the advanced features.

**Comment by nicorikken** [comment]
- The change to share the Asciidoctor JRuby container is committed.
- I tried to handle the YAML in Asciidoc, but could not manage to. So the only option would be to skip it altogether. Also I think keeping the same pattern as the Markdown task benefits the project structure. I consider exposing YAML metadata to the Asciidoctor tasks a desirable next step.
- I asked about the tests, which do not seem to test on content. Perhaps this can be improved at a later stage.

To me the code is ready to merge in it's current state. Metadata would be the focus for the next iteration.

**Comment by bhagany** [comment]
Seems to already be decided, but because this task can parse different types of content, it seems to be closer to `pandoc` (which is named after the tool), than it is to `markdown`.  So, I agree that `asciidoctor` is a more appropriate name for the task.

I do want to take a look at the tests, and I should have time tonight or tomorrow, but otherwise I agree this is ready to merge.  I'm waiting for the legal team at my new job to sign off on my work on Perun, so I don't want to merge before that happens.  If one of the other committers wants to after we figure out what's up with the tests though, I approve.

**Comment by bhagany** [comment]
Using aliases based on the type of content to be parsed is an interesting idea, and I think it deserves more thought.  There would be quite a few aliases for `pandoc`, but we'd probably be able to at least do the most popular ones.

**Comment by nicorikken** [comment]
@bhagany Thanks for the update. If somebody can verify I implemented the tests correctly, that would be great.

**Comment by bhagany** [comment]
@nicorikken I think the tests are implemented correctly.  I was able to get them to fail by changing the `parsed-asciidoc-md` and `parsed-asciidoc-adoc` vars to something else.  I might be missing something though, could you explain in more detail how you got them to behave in a way you didn't expect?

**Comment by nicorikken** [comment]
I was expecting a summary of the test results, as I know from my Leiningen usage. But indeed I get error output if the testresults are incorrect:
```
[asciidoctor] - rendered new or changed file test.html

FAIL in clojure.lang.PersistentList$EmptyList@1 (perun_test.clj:59)
io.perun-test/with-arguments-test
`asciidoctor` should populate HTML with parsed markdown content
expected: (f (.contains (slurp (boot/tmp-file (boot/tmp-get fileset path))) content))
  actual: false
```
So in hindsight it was just my expectation, not the actual code.

**Comment by bhagany** [comment]
Okay, then I think this is ready to merge.  Still haven't heard back from my legal department, but if another committer wants to merge before that happens, I'm good with that.

**Review by MartyGentillon — APPROVED** [review]

**Comment by nicorikken** [comment]
Considering that all participants on this PR approved, can this be merged?

**Comment by bhagany** [comment]
I got permission from the legal department :)

**Comment by nicorikken** [comment]
Nice!
Finally getting Asciidoctor is a relief. Thanks!

**Pull request merged**
